engine/crates/runtime-local/src/redis.rs

```diff
@@ -0,0 +1,132 @@
+mod pool;
+
+use std::{
+    collections::{hash_map::Entry, HashMap},
+    fs::File,
+    io::{BufReader, Read},
+    path::{Path, PathBuf},
+    time::Duration,
+};
+
+use anyhow::Context;
+use grafbase_telemetry::span::GRAFBASE_TARGET;
+use redis::ClientTlsConfig;
+
+pub type Pool = deadpool::managed::Pool<pool::Manager>;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct RedisTlsConfig<'a> {
+    pub cert: Option<&'a Path>,
+    pub key: Option<&'a Path>,
+    pub ca: Option<&'a Path>,
+}
+
+#[derive(PartialEq, Eq, Hash, Default)]
+struct RedisConfigKey {
+    url: String,
+    cert: Option<PathBuf>,
+    key: Option<PathBuf>,
+    ca: Option<PathBuf>,
+}
+
+/// A deduplicating factory for redis connection pools.
+///
+/// If you ask it to create a pool with the same details twice it will return the same pool
+#[derive(Default)]
+pub struct RedisPoolFactory {
+    pools: HashMap<RedisConfigKey, Pool>,
+}
+
+impl RedisPoolFactory {
+    pub fn pool(&mut self, url: &str, tls_config: Option<RedisTlsConfig<'_>>) -> anyhow::Result<Pool> {
+        let key = {
+            let mut config_key = RedisConfigKey {
+                url: url.to_string(),
+                ..Default::default()
+            };
+            if let Some(RedisTlsConfig { cert, key, ca }) = tls_config {
+                config_key = RedisConfigKey {
+                    cert: cert.map(ToOwned::to_owned),
+                    key: key.map(ToOwned::to_owned),
+                    ca: ca.map(ToOwned::to_owned),
+                    ..config_key
+                };
+            }
+            config_key
+        };
+
+        match self.pools.entry(key) {
+            Entry::Occupied(entry) => Ok(entry.get().clone()),
+            Entry::Vacant(entry) => {
+                let pool = new_pool(url, tls_config)?;
+                entry.insert(pool.clone());
+                Ok(pool)
+            }
+        }
+    }
+}
+
+fn new_pool(url: &str, tls_config: Option<RedisTlsConfig<'_>>) -> anyhow::Result<Pool> {
+    let tls_config = match tls_config {
+        Some(tls) => {
+            let client_tls = match tls.cert.zip(tls.key) {
```

Comment:
:zap: **Suggestion** 
In the `new_pool` function, consider handling the case where `tls.cert` or `tls.key` is `None` more explicitly, to avoid potential panics when calling `zip` on `None` values.